design-documents/frontend/technical-vision.md

```diff
@@ -0,0 +1,140 @@
+# Front-End Technical Vision
+
+The following document is a guide for all _future_ Magento front-end applications outside of core, for which we should be mindful of the following values:
+
+- Portability
+- Extensibility
+- Maintainability
+- Consistency
+- Performance
+- Accessibility
+- Internationalization
+- Productivity (Efficiency)
+- Quality (Reliability)
+
+## [React](https://reactjs.org/)
+
+_Portability, Performance, Productivity, Quality, Internationalization._
+
+React's [continued growth and popularity](https://www.npmtrends.com/@angular/core-vs-react-vs-vue) coupled with its [solid documentation](https://reactjs.org/docs/getting-started.html) make it an excellent choice for building user interfaces. If written appropriately, it enables **portability** by [crossing over into native applications](http://jkaufman.io/react-web-native-codesharing/) with [React Native](https://facebook.github.io/react-native/).
+
+**Performance** is easy to achieve with React, because it only re-renders DOM elements where and when necessary. That's not to say that React prevents you from creating **performance** problems, but the docs do well at explaining [what to avoid and how to profile](https://reactjs.org/docs/optimizing-performance.html).
+
+Creating a new React project has never been easier with [CRA][], which comes out-of-the-box with [hot reloading](https://webpack.js.org/concepts/hot-module-replacement/). This increases **productivity** by giving developers [short, positive feedback loops](https://www.youtube.com/watch?v=avKBrgIQXzk). Hot reloading is [also achievable in React Native](https://facebook.github.io/react-native/blog/2016/03/24/introducing-hot-reloading.html). Another opportunity for short, positive feedback loops is testing with [Jest][] and [Enzyme](https://airbnb.io/enzyme/). Tests are fastest when writing unit tests with the [Shallow Rendering API](https://airbnb.io/enzyme/docs/api/shallow.html) and reserving the [Full Rendering API](https://airbnb.io/enzyme/docs/api/mount.html) for integration tests. The better the development experience in these cycles, the better **quality** we achieve as a result. [Jest's snapshot testing](https://jestjs.io/docs/en/snapshot-testing) is another way to ensure **quality** in that we don't inadvertently introduce breaking changes.
+
+### Technologies
+- Create a new React project with [CRA][].
+- [Redux](https://redux.js.org/) as a predictable state container.
+   - Handling of data fetching and side effects now have first class support with [React Suspense, React.lazy](https://reactjs.org/blog/2018/10/23/react-v-16-6.html#reactlazy-code-splitting-with-suspense) and [hooks (proposal)](https://reactjs.org/docs/hooks-intro.html).
+   - Logging with [redux-logger](https://github.com/evgenyrodionov/redux-logger).
+- **Internationalization** with [react-intl](https://www.npmjs.com/package/react-intl).

```

Comment:
There are a _massive_ number of options in this space, some specific to React and some a bit more abstract. What's the motivation for recommending this particular lib?

Reply:
I didn't want to litter this document with all the plethora of choices available. Instead, this document is a guide. Here's a viable solution that we already use and has worked for us. Maybe you find something better in the future or even today that warrants a discussion point against this technology choice. By all means, present that choice as a better alternative to react-intl and defend it with argument points and we can change this item to something else.

This document doesn't have to be 🔐. I see it as a living document that:
1. Highlights or implies what we want (the general requirement).
1. Provides a viable solution to that requirement so you can hit the ground running without going on a technology hunt every time you need a technology in your application.

From there, I'm open to critical feedback against any proposed technology, so long as you pitch a good alternative.